### DISCUSSION TITLE
How to unzip a file?

### DISCUSSION THREAD

**Discussion by Androz2091** [101arrowz/fflate#44] [Q&A] [answered]
Hello :wave:

I guess I'm doing something wrong but I copied the code you sent me [here](https://github.com/Stuk/jszip/issues/749) and it doesn't work.

Here is my code:
```js
uz.onfile = (f) => {
    console.log(f); // this prints the file object successfully
    if (f.name == '') {
        const decoder = new DecodeUTF8();
        f.ondata = (err, data, final) => decoder.push(data, final);
        decoder.ondata = (str, final) => {
            // this throws the following error:

/*
browser.js:1532 Uncaught (in promise) TypeError: Cannot read property 'length' of undefined
    at DecodeUTF8.push (browser.js:1532)
    at Object.f.ondata (Loader.svelte:20)
    at AsyncUnzipInflate.d.ondata (browser.js:2263)
    at Inflate.ondata (browser.js:2162)
    at Inflate.c (browser.js:1004)
    at Inflate.push (browser.js:1014)
    at AsyncUnzipInflate.push (browser.js:2175)
    at Unzip.push (browser.js:2211)
    at Unzip.push (browser.js:2311)
    at Unzip.push (browser.js:2216)
*/

        }
        f.start();
    }
}

const reader = file.stream().getReader();
while (true) {
    const { done, value } = await reader.read();
    if (done) {
        uz.push(new Uint8Array(0), true);
        break;
    }
    uz.push(value); // this throws an the following error:

/*
Loader.svelte:35 Uncaught (in promise) RangeError: Maximum call stack size exceeded
    at strFromU8 (browser.js:1618)
    at _loop_2 (browser.js:2242)
    at Unzip.push (browser.js:2298)
    at Unzip.push (browser.js:2311)
    at Unzip.push (browser.js:2216)
    at Unzip.push (browser.js:2311)
    at Unzip.push (browser.js:2216)
    at Unzip.push (browser.js:2311)
    at Unzip.push (browser.js:2311)
    at Unzip.push (browser.js:2216) 
*/
}
```
**file** is an object from a file input.

Thank you

**Comment by 101arrowz** [comment, marked as answer]
Sorry, this was a bug. I've published v0.6.5 which fixes the first bug, but I haven't experienced the second one. Can you try the following code? If it throws no error, try your previous code and if it still fails, send me your ZIP file. If this code throws an error, tell me what browser, version, OS, etc. you are running.

```js
const uz = new Unzip();
uz.register(AsyncUnzipInflate)
const file = (await fetch('https://raw.githubusercontent.com/mrdoob/three.js/dev/examples/models/3mf/truck.3mf')).body;
uz.onfile = (f) => {
    console.log(f); // this prints the file object successfully
    if (f.name != '3D/3dmodel.model' && !f.name.includes('.png')) {
        const decoder = new DecodeUTF8();
        f.ondata = (err, data, final) => decoder.push(data, final)
        decoder.ondata = (str, final) => {
            console.log(f.name, str);
        }
        f.start();
    }
}

const reader = file.getReader();
while (true) {
    const { done, value } = await reader.read();
    if (done) {
        uz.push(new Uint8Array(0), true);
        break;
    }
    uz.push(value);
}
```
Side note, I noticed you're running the unminified build. Assuming you're using a CDN, make sure to use the CDN instructions in the README when running for production because those offer much less bundle size.

If you want to get the app working ASAP, you can try using the JSZip instructions I gave you, though performance should be an order of magnitude faster with `fflate`.

**Reply by Androz2091** [reply]
You are right, the first bug has been fixed when bumping the package! And I'm not using a CDN, but a JavaScript framework (Svelte) so I'm just importing the package using `import`. I'm still getting the other error, `Maximum call stack size exceeded`. But, the ZIP file I'm trying to read is my Discord Data Package (with all my messages, servers, etc. all my data) so I can't send it 😬

**Reply by 101arrowz** [reply]
In that case, try this on the original zip:

```js
import { unzip, DecodeUTF8 } from 'fflate';
unzip(new Uint8Array(await file.arrayBuffer()), (err, archive) => {
  const buf = archive['yourfile.json'];
  const decoder = new DecodeUTF8();
  decoder.ondata = (str, final) => console.log('streamed in:', str);
  for (let i = 0; i < file.size; i += 512000) {
    decoder.push(buf.subarray(i, i + 512000), i + 512000 >= file.size);
  }
})
```

`fflate`'s streaming ZIP decompression has had a few bugs. My guess is that if you unzip and rezip the archive, your original program wil work. However, you can use this too, it will be as fast as the other one but take more memory (your entire ZIP will need to be in memory, so around 200MB + 200MB used by fflate, whereas in the other one it's streamed in so only 64kB + 1MB used by fflate).

**Reply by Androz2091** [reply]
It works! But I used some `console.time`, and here are the results:
```js
console.time('buf')
const buf = await file.arrayBuffer()
console.timeEnd('buf')
console.time('unzip')
unzip(new Uint8Array(buf), (err, archive) => {
    console.timeEnd('unzip')
    console.time('read')
    const buf = archive['activity/analytics/events-2021-00000-of-00001.json'];
    const decoder = new DecodeUTF8();
    decoder.ondata = (str, final) => {
        if (final) console.timeEnd('read')
    }
    for (let i = 0; i < file.size; i += 512000) {
        decoder.push(buf.subarray(i, i + 512000), i + 512000 >= file.size);
    }
})
```
```js
buf: 82.987060546875 ms
unzip: 4888.302978515625 ms
read: 726.35888671875 ms
```
so the unzip function takes ~5 seconds to load, and when I use this code:
```js
console.time('unzip');
jszip.loadAsync(file).then((zip) => {
    console.timeEnd('unzip');
    // then do something else, like reading the files, etc.
```
I get `unzip: 441.115966796875 ms`. Do you know what could explain this difference?
Thank you

On a side note, here is my goal. I have a big data zip file (`200mb` when zipped, `1.90Go` when unzipped). In this file, there are three folders. The `messages` one contains approximately 3000 folders for 5.5mb of data (unzipped). I need to loop on these folders, and read the two files in them, a small `.json` file and a `.csv` file. So it represents 6,000 files to read. Reading these files is pretty fast using JSZip (`1152ms`). In the `user` folder, there is a small `.json` file so it's easy to read. But in the `analytics` folder... there is `900mb` `.json` file containing all the user statistics. And this is file is very hard to read. It is an array of events. Here is an example of event:
```json
{"event_type":"permissions_requested","event_id":"XXXXXX","user_id":"422820341791064085","domain":"Analytics","freight_hostname":"analytics-ingest-prd-7341","ip":"XXXXXX","day":"794","chosen_locale":"fr","detected_locale":"en-US","user_is_authenticated":true,"browser":"Discord Client","cfduid":"XXXXXX","os":"Mac OS X","os_version":"18.7.0","os_arch":"x64","client_build_number":"60259","release_channel":"stable","client_version":"0.0.257","client_performance_cpu":33.326694676970142,"client_performance_memory":"0","country_code":"FR","time_zone":"Europe/Paris","isp":"Orange","type":"audio","accessibility_support_enabled":false,"accessibility_features":"256","client_send_timestamp":"\"2020-05-15T11:30:17.865Z\"","client_track_timestamp":"\"2020-05-15T11:30:12.134Z\"","timestamp":"\"2020-05-15T11:30:18.134Z\""}
```
So the best would be to get a big JSON so I can generate interesting statistics like:
```js
const permissionsOnMacOS = events.filter((event) => event.event_type === 'permissions_requested' && event.os === 'Mac OS X')
```
But I guess it's not possible, as importing a `900mb` JSON in the memory in the browser is not the best idea... So I can still generate statistics by counting the number of occurrences of a certain string like `permissions_requested`. I could use a stream like the one you sent above and in each chunk I increment the `permissionsRequestedCount` variable if there is `permissions_requested` in the string. It's not the best but at least I can get some statistics... Do you have another idea?

**Reply by 101arrowz** [reply]
The reason JSZip is so much faster is that it doesn't actually decompress the files. `fflate` decompresses every file in the archive by default when you use `unzip`, but you can choose what to decompress when using `Unzip`. If you actually try to decompress every file in the archive, JSZip is *much* slower (about 50% from my test):

```js
const buf = await file.arrayBuffer()
console.time('jszip');
JSZip.loadAsync(buf).then(async zip => {
  for (const file in zip.files) {
    console.log(file, await zip.files[file].async('uint8array'));
  }
  console.timeEnd('jszip'); // 141.20ms for truck.3mf
});
console.time('fflate');
// For small files, this is faster
console.log(unzipSync(new Uint8Array(buf)));
console.timeEnd('fflate'); // 81.37ms for truck.3mf
// If your ZIP is very large and has multiple large files
unzip(new Uint8Array(buf), (err, archive) => {
  console.log(archive);
  console.timeEnd('fflate'); // slow on truck.3mf, over 3x faster on this:
  // http://imagecompression.info/test_images/
});
```

For the small files, use `const out = JSON.parse(strFromU8(archive['yourfile.json']))`. For the bigger files, you must use a streaming JSON parser (such as [this one](https://github.com/dscape/clarinet)) in combination with `DecodeUTF8`.

It's unfortunate I can't help you debug the `Unzip` version. Maybe try reducing the block size?
```js
const reader = file.stream().getReader();
while (true) {
    const { done, value } = await reader.read();
    if (done) {
        uz.push(new Uint8Array(0), true);
        break;
    }
    for (let i = 0; i < value.length; i += 65536) {
      uz.push(value.subarray(i, i + 65536));
    }
}
```

**Reply by Androz2091** [reply]
> The reason JSZip is so much faster is that it doesn't actually decompress the files. fflate decompresses every file in the archive by default when you use unzip, but you can choose what to decompress when using Unzip. If you actually try to decompress every file in the archive, JSZip is much slower (about 50% from my test):

All right I understand now

> Maybe try reducing the block size?

Woaw, it just... worked! :grin:
Yes it works perfectly now!
So I should use this way now right? And I'd like to understand what's the best way to read my files. Should I use something like this?
```js
if (f.name === 'account/user.json') {
    const decoder = new DecodeUTF8();
    const fileContent = [];
    f.ondata = (err, data, final) => decoder.push(data, final);
    decoder.ondata = (str, final) => {
        fileContent.push(str);
        if (final) {
            const data = JSON.parse(fileContent.join(''));
            console.log(data)
            console.timeEnd('read account');
        }
    }
    f.start();
}
```
Or should I create a `files` array (and push the file into in the onfile function) and then create a `readFile` method like:
```js
const readFile = (f) => {
    return new Promise((resolve) => {
        const decoder = new DecodeUTF8();
        const fileContent = [];
        f.ondata = (err, data, final) => decoder.push(data, final);
        decoder.ondata = (str, final) => {
            fileContent.push(str);
            if (final) resolve(fileContent);
        }
        f.start();
    });
}

const [user, someFiles ] = await Promise.all([
    readFile(files['account/user.json']),
    readFile(files['someotherfile.json'])
]);
```

And yes, `clarinet` seems very interesting 🤔 
but seems like I can't use it in the browser? as it uses stream

**Reply by Androz2091** [reply]
All right so here are my benchmarks for my 900mb file using fflate and clarinet.

### fflate + Clarinet

Read the file + JSON parsing using Clarinet

| Mobile Device | Computer |
------|-----
| 58 seconds | 14 seconds

It takes 58 seconds to read the file **on a mobile device** and 14 seconds **on a computer**

### Only fflate

Read the file only

| Mobile Device | Computer |
------|-----
| 15 seconds | 2 seconds

So it is very fast using fflate only. Do you have an idea for how to optimize the JSON parsing? And my file contains "permissions_requested" 1800 times. When using something like:
```js
uz.onfile = (f) => {
    if (f.name === 'activity/analytics/events-2021-00000-of-00001.json') {
        console.time('read analytics');
        console.time('read file');
        const decoder = new DecodeUTF8();
        let permRequested = 0;
        f.ondata = (err, data, final) => decoder.push(data, final);
        decoder.ondata = (str, final) => {
            if (str.includes('permissions_requested')) permRequested++;
            if (final) {
                console.timeEnd('read file');
                console.log(permRequested) // 7, 6 or sometimes 5
            }
        }
        f.start();
    }
}
```
As you can see, the number of occurences found is not correct at all, and is not constant. Do you have an idea why?

**Reply by 101arrowz** [reply]
The fact that it worked indicates that you have a metric ton of very tiny files/folders. You'd need over 11000 individual files and/or folders per megabyte of ZIP for the original code I gave you to crash, after subtracting headers this is an average of 60 bytes per file. I'll add a note to the README about this.

It's unfortunate that `clarinet` is so slow. I couldn't find any other JSON streaming parser that claimed to work at all in the browser, so  you'll just either have to deal with the performance loss, or not parse the JSON and instead just use `.includes`.

With regards to the issue you found, there are two problems with the code
1. Big problem - you increment if permissions_requested is found but you don't actually count the number of times the substring appears. Since you're looking for performance, use this code to count occurrences:
```js
let occurrences = 0;
const substr = 'permissions_requested';
while (true) {
  const ind = str.indexOf(substr);
  if (ind == -1) break;
  str = str.slice(ind + substr.length);
  occurrences++;
}
```
2. You aren't accounting for the case `permissions_requested` is on a boundary between chunks. For instance, one chunk ends with `"permissions_req` and the next chunk starts `uested": "the_data"`. So you need to keep a temp buffer that includes the last few characters of the previous chunk and concatenate. Update the code to this:
```js
let prevChkEnd = '';
decoder.ondata = (str, final) => {
  str = prevChkEnd + str;
  let occurrences = 0;
  const substr = 'permissions_requested';
  while (true) {
    const ind = str.indexOf(substr);
    if (ind == -1) break;
    str = str.slice(ind + substr.length);
    occurrences++;
  }
  prevChkEnd = str.slice(-substr.length);
  permRequested += occurrences;
}
```

**Reply by Androz2091** [reply]
Indeed, my bad, a chunk includes multiple events... Thank you 😅 
And all right, I will see for the json parser. And do you think there is a way to improve clarinet? Or we can definitely not parse json faster?

And is there a way to get a promise resolved where all the files have been retrieved by fflate?
I think I will use something like:
```js
const files = [];
 uz.onfile = (f) => {
    files.push(f);
    console.log(files.length)
}

extractData(files); // extractData is a function that will process the files, parse them and generate statistics. But I need to wait for all the files being retrieved by uz.onfile before calling it
```

And one more question 😅 
is there any way to get the progress of the parsing of a file? I saw #24 but don't think it's the same usage. Something like that for example:
```js
const decoder = new DecodeUTF8();
f.ondata = (err, data, final) => decoder.push(data, final);
decoder.ondata = (str, final, percent) => {
    console.log(percent); // 46
}
f.start();
```

**Reply by 101arrowz** [reply]
The reason true `JSON.parse` is so fast is because it calls an extremely optimized native function in the V8 engine (used in Chromium browsers). `clarinet` is slower because it's pure JavaScript. Of course, I'm a bit disappointed that a JSON parser is less optimized than a compression library, which has to do so many more things, but oh well...maybe I'll write one at some point, just because I have faced some similar annoyances with regards to large JSON files.

The `.push()` function synchronously finds all files in the archive; only decompression is optionally asynchronous. When `done` is `true`, you `break` from the loop that keeps pushing chunks, so right after that loop you can assume all files have been found. No need to wait for a Promise. (By the way, I'm sure you know this approach is a bit slower, but it's still much better than decompressing all the files at once).

```js
const reader = file.stream().getReader();
while (true) {
    const { done, value } = await reader.read();
    if (done) {
        uz.push(new Uint8Array(0), true);
        break;
    }
    for (let i = 0; i < value.length; i += 65536) {
      uz.push(value.subarray(i, i + 65536));
    }
}
// Here all files have been found
extractData(files);
```

You're right, #24 refers to progress for the entire ZIP file. If you want progress on a file-specific level, you might be able to get its original size and divide the number of bytes read to calculate a percent. However, sometimes this is not possible because the ZIP archive does not encode data about the length of each file in the header.

```js
// This is undefined for some ZIP files, unfortunately
// In those cases percentage is impossible to get
const origSize = f.originalSize;
let bytesRead = 0;
const decoder = new DecodeUTF8();
f.ondata = (err, data, final) => {
  if (origSize) {
    bytesRead += data.length;
    console.log(`${bytesRead / origSize * 100}%`);
  }
  decoder.push(data, final);
}
decoder.ondata = (str, final) => {
  // logic here
}
f.start();
```

**Reply by Androz2091** [reply]
`f.originalSize` is defined in my package and it works perfectly! Thank you so much for your help, I'm going to use your library for https://github.com/Androz2091/discord-data-package-explorer

If you build your own json parser library you can email me at androz2091@gmail.com, it would be very useful for my website!

**Reply by Androz2091** [reply]
Hmm... I have another problem (I mean, maybe it's just me but I don't see why so I think it is related to the library). This is pretty weird

```js
const readFile = (fileToRead) => {
    const fileContent = [];
    const decoder = new DecodeUTF8();
    fileToRead.ondata = (err, data, final) => {
        decoder.push(data, final);
    };
    decoder.ondata = (str, final) => {
        fileContent.push(str);
        if (final) console.log(fileContent.join(''));
    };
    fileToRead.start()
}

const uz = new Unzip();
uz.register(AsyncUnzipInflate);

const files = [];
uz.onfile = (f) => {
    files.push(f);
    if (f.name === 'account/user.json') readFile(f); // prints the file successfully in the console
}

const reader = file.stream().getReader();
while (true) {
    const { done, value } = await reader.read();
    if (done) {
        uz.push(new Uint8Array(0), true);
        break;
    }
    for (let i = 0; i < value.length; i += 65536) {
        uz.push(value.subarray(i, i + 65536));
    }
}

const userFile = files.find((file) => file.name === 'account/user.json');
readFile(userFile); // throws the following error:
/*
browser.js:2281 Uncaught (in promise) TypeError: Cannot read property '0' of null
at Object.start (browser.js:2281)
at readFile (Loader.svelte:23)
at handleFile (Loader.svelte:47)
*/
```
it would mean that I can only read file in the `onfile` method? Weird, right?

**Reply by 101arrowz** [reply]
Funny enough, this was a reversion caused by publishing v0.6.5. It is intended to be able to use `.start()` outside of `onfile`. I've fixed the reversion in v0.6.6, everything should work now (fingers crossed).

**Reply by Androz2091** [reply]
All right, it works now 👍 
I'm facing another issue, for one specific file `final` is never `true` (there is no more data passed to the file.ondata but `final` is still false). Do you have any idea why? (`err` is `null` when receiving the last chunk - actually the only chunk because there is only one chunk with this file)

Tell me if you want me to open issues instead of reporting here

**Reply by 101arrowz** [reply]
Just keep it here, prevents clogging the issues page with these one line changes. Fixed this bug in v0.6.7. Sorry for all the problems! You're the first person to use streaming unzip on a very large file, so I'm trying to figure things out.

**Reply by Androz2091** [reply]
All right. and thank you it works now 👍
